packages/lib/pod/src/podJSON.ts

```diff
@@ -0,0 +1,345 @@
+export function podEntriesToJSON(podEntries: PODEntries): JSONPODEntries {
+  requireType("podEntries", podEntries, "object");
+  return Object.fromEntries(
+    Object.entries(podEntries).map(([name, podValue]) => [
+      checkPODName(name),
+      podValueToJSON(podValue, name)
+    ])
+  );
+}
+
+/**
+ * Converts a {@link PODValue} from the TypeScript format used internally, into
+ * a JSON-compatible format safe to serialize with `JSON.stringify`.  This
+ * function will produce the simplest of the available encodings which can
+ * represent the input with no loss value or type information.
+ *
+ * @param podValue the POD entries to convert
+ * @param nameForErrorMessages an optional name for this value to be used
+ *   in error messages
+ * @returns a JSON-compatible representation
+ * @throws TypeError if the input entries are not validly formed
+ */
+export function podValueToJSON(
+  podValue: PODValue,
+  nameForErrorMessages?: string
+): JSONPODValue {
+  nameForErrorMessages = nameForErrorMessages || "(unnamed)";
+  podValue = checkPODValue(nameForErrorMessages, podValue);
+  switch (podValue.type) {
+    case "string":
+      return podValue.value;
+    case "int":
+      const numValue = bigintToSimplestJSON(podValue.value);
+      if (typeof numValue === "number") {
+        return numValue;
+      }
+      return { int: numValue };
+    case "cryptographic":
+      return { cryptographic: bigintToSimplestJSON(podValue.value) };
+    case "eddsa_pubkey":
+      return { eddsa_pubkey: podValue.value };
+    default:
+      throw TypeError(
+        `Value ${nameForErrorMessages} has unhandled POD value ` +
+          // @ts-expect-error podValue is of type `never` if we've covered all types
+          `type '${podValue.type}'.`
+      );
+  }
+}
+
+/**
+ * Converts a bigint to the simplest value-preserving JSON value available.
+ * There is no type information included.  The output is always a number
+ * (with no fractional part) or a string which can be parsed by `BigInt(s)`.
+ *
+ * Numbers are used preferentially for values in the range between
+ * `Number.MIN_SAFE_INTEGER` and `Number.MAX_SAFE_INTEGER` where no accuracy
+ * is lost.  Stringified values may be hex or decimal encoded for minimal size.
+ *
+ * Note that this function doesn't enforce any bounds on the input.  Any
+ * bigint will be encoded, even if it's not a valid POD value.
+ *
+ * @param n the input integer
+ * @returns a number or string representing this integer.
+ */
+export function bigintToSimplestJSON(n: bigint): number | string {
+  // For values which are in range for a JSON number (2^53), use that.
+  if (n <= Number.MAX_SAFE_INTEGER && n >= Number.MIN_SAFE_INTEGER) {
+    return Number(n);
+  }
+
+  // Large positive values are shorter as hex than decimal.
+  // MAX_SAFE_INTEGER is large enough to be equal size even with the addition of
+  // "0x", and the benefits become greater for larger values.
+  if (n > 0) {
+    return "0x" + n.toString(16);
```

Comment:
This trades off readability for some terseness, but I'm not sure if this is worth it. I'd be a bit surprised to find some of my numbers (but only _some_ of them, for reasons that are far from obvious to the casual observer) turning into hexadecimal representations. I'm imagining someone grepping the logs of data received from HTTP POST requests that might contain a specific user's Semaphore commitment, and searching for the decimal representation and not finding it, because that number is big enough to be turned into a hexadecimal number, while other numbers they have seen previously were not.

It also relies on an implementation detail of the JavaScript `bigint` constructor to be able to cope with both decimal and hex string representations of the same number, which a naive implementer of a non-JavaScript version of this format might easily be caught out by.

Reply:
Discussed offline.  We're both on the fence on this point, but decided to keep it in.
We also noted that there are other formats accepted by the BigInt constructor, such as "0o123" for Octal, which we'll never produce, but could accept if someone decided to use it.